File: 3차/3주차/파괴되지 않은 건물.cpp
#include <iostream>
#include <vector>
using namespace std;
#define MAX 1005
int Sum[MAX][MAX];
int solution(vector<vector<int>>board, vector<vector<int>> skill) {
    int answer = 0;
    int sk[6]; // type,r1,c1,r2,c2,degree;
    for (auto i : skill)
    {
        for (int j = 0; j < 6; j++)
            sk[j] = i[j];
        int type = sk[0];
        int r1 = sk[1];
        int c1 = sk[2];
        int r2 = sk[3];
        int c2 = sk[4];
        int degree = sk[5];
        if (type == 1) degree = -degree;
        Sum[r1][c1] += degree;
        Sum[r1][c2 + 1] -= degree;
        Sum[r2 + 1][c1] -= degree;
        Sum[r2 + 1][c2 + 1] += degree;
    }
    int n = board.size();
    int m = board[0].size();
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            Sum[i][j + 1] += Sum[i][j];
    for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
            Sum[j + 1][i] += Sum[j][i];
    for (int i = 0; i < board.size(); i++)
        for (int j = 0; j < board[i].size(); j++)
        {
            if (board[i][j] + Sum[i][j] > 0)
                answer++;
        }
    return answer;
}

File: 3차/3주차/고주희/파괴되지않은건물.java
class Solution {
    public int solution(int[][] board, int[][] skill) {
        int N = board.length;
        int M = board[0].length;
        int[][] tmp = new int[N][M];
        for (int[] s : skill) {
            int type = s[0];
            int r1 = s[1];
            int c1 = s[2];
            int r2 = s[3];
            int c2 = s[4];
            int degree = s[5];
            
            if(type == 1) degree *= -1;
            
            tmp[r1][c1] += degree;
            if(c2 < M - 1) tmp[r1][c2 + 1] -= degree;
            if(r2 < N - 1) tmp[r2 + 1][c1] -= degree;
            if(c2 < M - 1 && r2 < N - 1) tmp[r2 + 1][c2 + 1] += degree;
        }
        int answer = 0;
        for (int i = 1; i < N; i++) {
            for (int j  = 0; j < M; j++) {
                tmp[i][j] += tmp[i - 1][j];
            }
        }
        for(int j = 1; j < M; j++) {
            for(int i = 0; i < N; i++) {
                tmp[i][j] += tmp[i][j - 1];
            }
        }
        
        for(int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if(tmp[i][j] + board[i][j] > 0) answer++;
            }
        }
        return answer;
    }
}
